### PR TITLE
Add grpc_endpoint_get_fd and grpc_transport_get_endpoint

### PR DESCRIPTION
**Pull request by y-zeng**
- `grpc_transport_get_endpoint` gets the endpoint used by the transport
- `grpc_endpoint_get_socket` gets the socket (fd in posix, SOCKET in windows) used by the endpoint. It returns NULL if the endpoint does not have a socket.
- Add a type `GRPC_SOCKET`, which is `int` in posix and `SOCKET` in windows.

### CODE REVIEW

**Comment by murgatroid99** [comment]
What do we need `grpc_endpoint_get_socket` for?

**Comment by y-zeng** [comment]
These two methods can be used by a channel filter to retrieve information about socket options.

File: src/core/lib/iomgr/endpoint.h [hunk 1]
```diff
@@ -73,6 +74,10 @@ void grpc_endpoint_read(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep,
 
 char *grpc_endpoint_get_peer(grpc_endpoint *ep);
 
+/* Get the file descriptor used by \a ep. Return -1 if \a ep is not using a fd.
```
Comment (dgquintas) [inline src/core/lib/iomgr/endpoint.h]: super nit: s/a fd/an fd

**Comment by y-zeng** [comment]
Test failures seem unrelated.

**Pull request merged**